File: backend/app/schemas/radar.py
from typing import Literal

from pydantic import BaseModel, Field


WidgetType = Literal[
    "genome_map",
    "saturation_chart",
    "opportunity_scorecard",
    "competitor_matrix",
    "creative_brief",
    "luma_concepts",
]


class RadarMetric(BaseModel):
    label: str
    value: str


class RadarAlert(BaseModel):
    level: str
    text: str


class RadarActionPayload(BaseModel):
    widget: WidgetType


class RadarAction(BaseModel):
    type: str
    payload: RadarActionPayload


class RadarEditSuggestion(BaseModel):
    id: str
    title: str
    description: str
    action: RadarAction


class RadarBrief(BaseModel):
    title: str
    narrative: str
    metrics: list[RadarMetric]
    alerts: list[RadarAlert]
    strategy: list[str]


class RadarBackendTrace(BaseModel):
    mode: str
    inputs: list[str]
    pipeline: list[str]
    assumptions: list[str]
    confidence: str
    output: str


class RadarResponse(BaseModel):
    text: str
    thinking: list[str]
    widget: WidgetType
    brief: RadarBrief
    backendTrace: RadarBackendTrace
    suggestions: list[str]
    editSuggestions: list[RadarEditSuggestion]


class RadarChatRequest(BaseModel):
    prompt: str = Field(min_length=1, max_length=4000)
    brand: str | None = None
    category: str | None = None
    meta_signals: list[str] = Field(default_factory=list)
    campaign_context: str | None = None


class RadarChatEnvelope(BaseModel):
    ok: bool
    source: Literal["meta_ad_library"]
    mode: Literal["live", "fallback"]
    result: RadarResponse

File: frontend/src/types/radar.ts
export type WidgetType =
  | "genome_map"
  | "saturation_chart"
  | "opportunity_scorecard"
  | "competitor_matrix"
  | "creative_brief"
  | "luma_concepts";

export interface RadarMetric {
  label: string;
  value: string;
}

export interface RadarAlert {
  level: string;
  text: string;
}

export interface RadarActionPayload {
  widget: WidgetType;
}

export interface RadarAction {
  type: string;
  payload: RadarActionPayload;
}

export interface RadarEditSuggestion {
  id: string;
  title: string;
  description: string;
  action: RadarAction;
}

export interface RadarBrief {
  title: string;
  narrative: string;
  metrics: RadarMetric[];
  alerts: RadarAlert[];
  strategy: string[];
}

export interface RadarBackendTrace {
  mode: string;
  inputs: string[];
  pipeline: string[];
  assumptions: string[];
  confidence: string;
  output: string;
}

export interface RadarResponse {
  text: string;
  thinking: string[];
  widget: WidgetType;
  brief: RadarBrief;
  backendTrace: RadarBackendTrace;
  suggestions: string[];
  editSuggestions: RadarEditSuggestion[];
}

export interface RadarChatRequest {
  prompt: string;
  brand?: string;
  category?: string;
  meta_signals?: string[];
  campaign_context?: string;
}

export interface RadarChatEnvelope {
  ok: boolean;
  source: string;
  mode: "live" | "fallback";
  result: RadarResponse;
}
